src/probability/upcrossing.lean

```diff
@@ -0,0 +1,914 @@
+end
+
+/-- **Doob's upcrossing estimate**: given a real valued discrete submartingale `f` and real
+values `a` and `b`, we have `(b - a) * 𝔼[upcrossings_before a b f N] ≤ 𝔼[(f N - a)⁺]` where
+`upcrossings_before a b f N` is the number of times the process `f` crossed from below `a` to above
+`b` before the time `N`. -/
+lemma submartingale.mul_integral_upcrossings_before_le_integral_pos_part [is_finite_measure μ]
+  (a b : ℝ) (hf : submartingale f ℱ μ) (N : ℕ) :
+  (b - a) * μ[upcrossings_before a b f N] ≤ μ[λ x, (f N x - a)⁺] :=
+begin
+  by_cases hab : a < b,
+  { exact mul_integral_upcrossings_before_le_integral_pos_part' hf hab },
+  { rw [not_lt, ← sub_nonpos] at hab,
+    exact le_trans (mul_nonpos_of_nonpos_of_nonneg hab (integral_nonneg (λ x, nat.cast_nonneg _)))
+      (integral_nonneg (λ x, lattice_ordered_comm_group.pos_nonneg _)) }
+end
+
+lemma upcrossings_before_eq_sum (hN : 0 < N) (hab : a < b) :
+  upcrossings_before a b f N x =
+  ∑ i in finset.Ico 1 (N + 1), {n | upper_crossing_time a b f N n x < N}.indicator 1 i :=
+begin
+  rw ← finset.sum_Ico_consecutive _ (nat.succ_le_succ zero_le')
+    (nat.succ_le_succ (upcrossings_before_le f x hN hab)),
+  have h₁ : ∀ k ∈ finset.Ico 1 (upcrossings_before a b f N x + 1),
+    {n : ℕ | upper_crossing_time a b f N n x < N}.indicator 1 k = 1,
+  { rintro k hk,
+    rw finset.mem_Ico at hk,
+    rw set.indicator_of_mem,
+    { refl },
+    { refine upper_crossing_time_lt_of_le_upcrossings_before hN hab (nat.lt_succ_iff.1 hk.2) } },
+  have h₂ : ∀ k ∈ finset.Ico (upcrossings_before a b f N x + 1) (N + 1),
+    {n : ℕ | upper_crossing_time a b f N n x < N}.indicator 1 k = 0,
+  { rintro k hk,
+    rw [finset.mem_Ico, nat.succ_le_iff] at hk,
+    rw set.indicator_of_not_mem,
+    simp only [set.mem_set_of_eq, not_lt],
+    exact (upper_crossing_time_eq_of_upcrossings_before_lt hab hk.1).symm.le },
+  rw [finset.sum_congr rfl h₁, finset.sum_congr rfl h₂, finset.sum_const, finset.sum_const,
+    smul_eq_mul, mul_one, smul_eq_mul, mul_zero, nat.card_Ico, nat.add_succ_sub_one,
+    add_zero, add_zero],
+end
+
+lemma adapted.measurable_upcrossings_before (hf : adapted ℱ f) (hab : a < b) :
+  measurable (upcrossings_before a b f N) :=
+begin
+  by_cases hN : N = 0,
+  { rw [hN, upcrossings_before_zero'],
+    exact measurable_zero },
+  { have : upcrossings_before a b f N =
+      λ x, ∑ i in finset.Ico 1 (N + 1), {n | upper_crossing_time a b f N n x < N}.indicator 1 i,
+    { ext x,
+      exact upcrossings_before_eq_sum (zero_lt_iff.2 hN) hab },
+    rw this,
+    exact finset.measurable_sum _ (λ i hi, measurable.indicator measurable_const $
+      ℱ.le N _ (hf.is_stopping_time_upper_crossing_time.measurable_set_lt_of_pred N)) },
+end
+
+lemma adapted.integrable_upcrossings_before [is_finite_measure μ]
+  (hf : adapted ℱ f) (hab : a < b) :
+  integrable (λ x, (upcrossings_before a b f N x : ℝ)) μ :=
+begin
```

Comment:
Your proof looks very complicated: you already know that this function is measurable and bounded, so you shouldn't need anything more.